Unable to save files on exFAT drive.

### Summary

I keep getting "file modified by an external process, use w! to overwrite" when I am working on a secondary exFAT drive. Weirdly, this does not happen on the primary NTFS drive. 

### Reproduction Steps

https://asciinema.org/a/EEWHjojbUnUA06gIAxQZ9PgyT

All I am trying to do is save a file. 

### Helix log

Relevant logs for the session. 
<details><summary>~/.cache/helix/helix.log</summary>

```
Caused by:
    channel closed)
2024-05-27T14:39:15.513 helix_lsp::transport [ERROR] rust-analyzer err: <- StreamClosed
2024-05-27T15:33:39.312 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(23)), original request likely timed out
2024-05-28T10:52:25.810 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(25)), original request likely timed out
2024-05-28T12:21:34.692 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.692499Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.694 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.692546Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.694 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.692593Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.694 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.692607Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.935 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.935449Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.935 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.935485Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.935 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.935501Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.935 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.935554Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.935 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.935569Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.935 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.935582Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.937 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.937155Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.937 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.937219Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.952 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.952178Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.952 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.952215Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.952 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.952270Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.952 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.952288Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.992 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.992114Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.992 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.992141Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.992 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.992174Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.992 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.992181Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\parser.rs\n"
2024-05-28T12:21:34.994 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.994838Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.994877Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.994952Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.994974Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.995036Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.995060Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.995126Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:21:34.995 helix_lsp::transport [ERROR] rust-analyzer err <- "2024-05-28T16:21:34.995146Z ERROR rust_analyzer::main_loop: flycheck 0: File with cargo diagnostic not found in VFS: file not found: d:\\repos\\lang\\src\\main.rs\n"
2024-05-28T12:22:40.081 helix_lsp::transport [ERROR] err: <- Other(failed to send a message to server

Caused by:
    channel closed)
2024-05-28T12:22:40.081 helix_lsp::transport [ERROR] rust-analyzer err: <- StreamClosed

```

</details>


### Platform

Windows

### Terminal Emulator

wezterm 20240203-110809-5046fc22

### Installation Method

scoop

### Helix Version

helix 24.3 (2cadec0b)

--- a/helix-term/src/application.rs
+++ b/helix-term/src/application.rs
@@ -574,7 +574,7 @@ impl Application {
             doc_save_event.revision
         );
 
-        doc.set_last_saved_revision(doc_save_event.revision);
+        doc.set_last_saved_revision(doc_save_event.revision, doc_save_event.save_time);
 
         let lines = doc_save_event.text.len_lines();
         let bytes = doc_save_event.text.len_bytes();

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -104,6 +104,7 @@ impl Serialize for Mode {
 #[derive(Debug, Clone)]
 pub struct DocumentSavedEvent {
     pub revision: usize,
+    pub save_time: SystemTime,
     pub doc_id: DocumentId,
     pub path: PathBuf,
     pub text: Rope,
@@ -960,6 +961,11 @@ impl Document {
             }
             .await;
 
+            let save_time = match fs::metadata(&write_path).await {
+                Ok(metadata) => metadata.modified().map_or(SystemTime::now(), |mtime| mtime),
+                Err(_) => SystemTime::now(),
+            };
+
             if let Some(backup) = backup {
                 if write_result.is_err() {
                     // restore backup
@@ -982,6 +988,7 @@ impl Document {
 
             let event = DocumentSavedEvent {
                 revision: current_rev,
+                save_time,
                 doc_id,
                 path,
                 text: text.clone(),
@@ -1040,6 +1047,26 @@ impl Document {
         }
     }
 
+    pub fn pickup_last_saved_time(&mut self) {
+        self.last_saved_time = match self.path.as_mut().unwrap().metadata() {
+            Ok(metadata) => match metadata.modified() {
+                Ok(mtime) => mtime,
+                Err(_) => {
+                    log::error!(
+                        "Use a system time instead of fs' mtime not supported on this platform"
+                    );
+                    SystemTime::now()
+                }
+            },
+            Err(e) => {
+                log::error!(
+                    "Use a system time instead of fs' mtime: failed to file's metadata: {e}"
+                );
+                SystemTime::now()
+            }
+        };
+    }
+
     // Detect if the file is readonly and change the readonly field if necessary (unix only)
     pub fn detect_readonly(&mut self) {
         // Allows setting the flag for files the user cannot modify, like root files
@@ -1077,9 +1104,7 @@ impl Document {
         self.apply(&transaction, view.id);
         self.append_changes_to_history(view);
         self.reset_modified();
-
-        self.last_saved_time = SystemTime::now();
-
+        self.pickup_last_saved_time();
         self.detect_indent_and_line_ending();
 
         match provider_registry.get_diff_base(&path) {
@@ -1118,6 +1143,7 @@ impl Document {
         self.path = path;
 
         self.detect_readonly();
+        self.pickup_last_saved_time();
     }
 
     /// Set the programming language for the file and load associated data (e.g. highlighting)
@@ -1581,7 +1607,7 @@ impl Document {
     }
 
     /// Set the document's latest saved revision to the given one.
-    pub fn set_last_saved_revision(&mut self, rev: usize) {
+    pub fn set_last_saved_revision(&mut self, rev: usize, save_time: SystemTime) {
         log::debug!(
             "doc {} revision updated {} -> {}",
             self.id,
@@ -1589,7 +1615,7 @@ impl Document {
             rev
         );
         self.last_saved_revision = rev;
-        self.last_saved_time = SystemTime::now();
+        self.last_saved_time = save_time;
     }
 
     /// Get the document's latest saved revision.

--- a/helix-view/src/editor.rs
+++ b/helix-view/src/editor.rs
@@ -2084,7 +2084,7 @@ impl Editor {
                 };
 
                 let doc = doc_mut!(self, &save_event.doc_id);
-                doc.set_last_saved_revision(save_event.revision);
+                doc.set_last_saved_revision(save_event.revision, save_event.save_time);
             }
         }
 

